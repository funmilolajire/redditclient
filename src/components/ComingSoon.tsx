import { useRef, useLayoutEffect } from 'react';
import { gsap } from 'gsap';

export const ComingSoon = () => {
    const boxRef = useRef<HTMLDivElement>(null);

    useLayoutEffect(() => {
        const tl = gsap.timeline({ repeatDelay: 2 })
        tl.from([boxRef.current], {
            y: '-55vh',
            duration: 3,
            delay: 2,
            opacity: 0,
            ease: 'elastic'
        })
    })

    return (
        <div ref={boxRef} className="container flex items-center justify-center h-screen font-extrabold ComingSoon">
            <div className="bg-yellow-500 border-2 border-gray-800 border-solid skew-y-2 transform-gpu text-gray-50 text-9xl">
                <p>Coming Soon</p>
            </div>
        </div>
    )
}