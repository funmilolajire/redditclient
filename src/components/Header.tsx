import { useRef, useLayoutEffect } from 'react';
import { gsap } from 'gsap';
import logo from '../images/logo192.png';

export const Header = () => {
    const boxRef = useRef<HTMLDivElement>(null);
    const logoRef = useRef<HTMLImageElement>(null);
    const titleRef = useRef<HTMLHeadingElement>(null);

    useLayoutEffect(() => {
        const tl = gsap.timeline({});
        tl.from([boxRef.current], {
            duration: 1
        })
        tl.from([logoRef.current], {
            x: '-3000px',
            opacity: 0,
            ease: 'circ'
        })
        tl.from([titleRef.current], {
            x: '3000px',
            opacity: 0,
            ease: 'slow'
        })
    })

    return (
        <header>
            <div className="header-inner">
                <section ref={boxRef}>
                    <figure ref={logoRef}>
                        <img src={logo} alt="site logo" />
                    </figure>
                    <h1 ref={titleRef}>
                        <span>Reddit</span>Client
                    </h1>
                </section>
            </div>
        </header>
    )
}