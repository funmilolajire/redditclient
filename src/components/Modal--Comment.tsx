import { useRef } from 'react';
import parse from 'html-react-parser';
import { BiComment } from 'react-icons/bi';
import { BsDot } from 'react-icons/bs';
import { GoArrowDown, GoArrowUp } from 'react-icons/go';

interface ModalCommentType {
    [key: string]: any
}

export const ModalComment = ({ comment }: { comment: ModalCommentType }) => {
    const modalCommentRef = useRef<HTMLDivElement>(null);

    //parse comment html
    const parsedHTML = (html: string) => html ? parse(parse(html).toString()) : null;

    //JSX Element
    return (
        <div className="ModalComment" ref={modalCommentRef} data-permalink={comment.permalink}>
            <p className="Comment--Info">comment made <span className="time">{comment.time}</span> by <span className="author">{comment.author}</span></p>
            <div className="Comment--Body">
                {parsedHTML(comment.comment)}
            </div>
            <div className="Comment--Bottom">
                <div className="Comment--Replies">
                    <BiComment className="reply-icon icon" />
                    <span className="comments value">{comment.num_replies}</span>
                    <span className="label">{comment.num_replies !== 1 ? 'replies' : 'reply'}</span>
                </div>
                <BsDot className="dot-icon" />
                <div className="Votes Comment--Upvotes">
                    <GoArrowUp className="arrow-icon icon" />
                    <span className="upvotes value">{comment.upvotes}</span>
                    <span className="label">{comment.upvotes !== 1 ? 'upvotes' : 'upvote'}</span>
                </div>
                <BsDot className="dot-icon" />
                <div className="Votes Comment--Downvotes">
                    <GoArrowDown className="arrow-icon icon" />
                    <span className="downvotes value">{comment.downvotes}</span>
                    <span className="label">{comment.downvotes !== 1 ? 'downvotes' : 'downvote'}</span>
                </div>
            </div>
        </div>
    )
}