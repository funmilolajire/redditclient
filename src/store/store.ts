import { configureStore } from '@reduxjs/toolkit';
import subredditsReducer from './slices/subredditsSlice';
import postsReducer from './slices/postsSlice';
import moreInformationReducer from './slices/moreInformationSlice';

export const store = configureStore({
  reducer: {
    subreddits: subredditsReducer,
    posts: postsReducer,
    moreInformation: moreInformationReducer
  },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch