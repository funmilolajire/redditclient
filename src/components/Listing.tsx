import { useRef, useLayoutEffect, useCallback } from 'react';
import { useAppDispatch, useAppSelector } from '../store/hooks';
import { handleListingChange, handleTermChange } from '../store/slices/postsSlice';

export const Listing = ({ listItem }: { listItem: string }) => {
    const listingRef = useRef<HTMLLIElement>(null);

    //handle listing change and active class
    const dispatch = useAppDispatch();
    const listing = useAppSelector(state => state.posts.listing);
    const handleListing = () => {
        if (null !== listingRef.current && listing !== listingRef.current.id) {
            let newListing = listingRef.current.id;
            dispatch(handleListingChange(newListing))
            dispatch(handleTermChange(''))
        }
    }
    const handleActiveClass = useCallback(() => {
        if (null !== listingRef.current) {
            const classList = listing === listingRef.current.id ? "Listing active" : "Listing";
            listingRef.current.className = classList
        }
    }, [listing])
    useLayoutEffect(() => {
        handleActiveClass();
    }, [handleActiveClass])


    //JSX
    return (
        <li ref={listingRef} data-testid={listItem} onClick={handleListing} id={listItem}>
            {listItem}
        </li>
    )
}