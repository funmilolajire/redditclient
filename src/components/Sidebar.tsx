import { useRef, useLayoutEffect, useEffect } from 'react';
import { Searchbar } from "./Searchbar";
import { Subreddits } from './Subreddits';
import { gsap } from 'gsap';
import { HiOutlineMenuAlt2 } from 'react-icons/hi';
import { IoCloseOutline } from 'react-icons/io5';
import { useAppSelector } from '../store/hooks';

export const Sidebar = () => {
    const boxRef = useRef<HTMLDivElement>(null);
    const menuRef = useRef<HTMLDivElement>(null);
    useLayoutEffect(() => {
        const tl = gsap.timeline({})
        tl.from([boxRef.current], {
            y: '-120%',
            delay: 0,
            duration: 2,
            ease: 'power3.inOut'
        })
        tl.from([menuRef.current], {
            y: '-120%',
            delay: 0,
            duration: 2,
            ease: 'power3.inOut'
        })
    }, [])

    const openMobileMenu = () => {
        const mobileMenu = menuRef.current?.dataset.mobileMenu
        if (mobileMenu) {
            menuRef.current?.classList.remove('sm:flex')
            boxRef.current?.classList.remove('sm:hidden')
        }
    }

    const closeMobileMenu = () => {
        const menu = boxRef.current?.dataset.menu
        if (menu) {
            boxRef.current?.classList.add('sm:hidden')
            menuRef.current?.classList.add('sm:flex')
        }
    }

    const subreddit = useAppSelector(state => state.posts.subreddit)
    const listing = useAppSelector(state => state.posts.listing);
    const term = useAppSelector(state => state.posts.term);
    const showModal = useAppSelector(state => state.moreInformation.showModal);

    useEffect(() => {
        closeMobileMenu()
    }, [subreddit, listing, term, showModal])

    return (
        <>
            <aside ref={boxRef} data-menu className="Sidebar sm:hidden">
                <div className="Sidebar--Container">
                    <span onClick={closeMobileMenu} className="close"><IoCloseOutline /> </span>
                    <Searchbar />
                    <Subreddits />
                </div>
            </aside>

            <aside ref={menuRef} data-mobile-menu onClick={openMobileMenu} className="Mobile--Sidebar sm:flex hidden">
                <span className="menu"><HiOutlineMenuAlt2 /> </span>
            </aside>
        </>
    )
}