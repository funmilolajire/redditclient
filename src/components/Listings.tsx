import { useRef, useLayoutEffect } from 'react';
import { Listing } from './Listing';
import { gsap } from 'gsap';

export const Listings = () => {
    const listings: string[] = ['hot', 'new', 'top', 'best', 'rising', 'controversial'];

    //animations
    const listingBoxRef = useRef<HTMLDivElement>(null);
    useLayoutEffect(() => {
        const tl = gsap.timeline({})
        tl.from([listingBoxRef.current], {
            x: '1200%',
            delay: 2,
            duration: 2,
            ease: 'power3.inOut'
        })
    }, [])

    return (
        <div ref={listingBoxRef} className="Listings">
            <ul className="Listings--List">
                {
                    listings && listings.map(listing => <Listing key={listing} listItem={listing} />)
                }
            </ul>
        </div>
    )
}