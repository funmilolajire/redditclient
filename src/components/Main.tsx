import { Listings } from './Listings';
import { Posts } from './Posts';
export const Main = () => (
    <div className="Main">
        <Listings />
        <Posts />
    </div>
)