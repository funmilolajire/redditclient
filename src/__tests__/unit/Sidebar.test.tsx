import { mount } from 'enzyme';
import { Sidebar } from '../../components/Sidebar';
import { Provider } from 'react-redux';
import { store } from '../../store/store';

describe('renders Sidebar component', () => {
    const SidebarComponent = mount(
        <Provider store={store}>
            <Sidebar />
        </Provider>
    )

    test('make a snapshot of Sidebar Component', () => {
        expect(SidebarComponent).toMatchSnapshot()
    })
})