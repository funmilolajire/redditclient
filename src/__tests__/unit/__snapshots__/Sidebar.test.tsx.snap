// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Sidebar component make a snapshot of Sidebar Component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Sidebar>
    <aside
      className="Sidebar sm:hidden"
      data-menu={true}
    >
      <div
        className="Sidebar--Container"
      >
        <span
          className="close"
          onClick={[Function]}
        >
          <IoCloseOutline>
            <IconBase
              attr={
                Object {
                  "viewBox": "0 0 512 512",
                }
              }
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M368 368L144 144m224 0L144 368"
                  fill="none"
                  key="0"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="32"
                />
              </svg>
            </IconBase>
          </IoCloseOutline>
           
        </span>
        <Searchbar>
          <div
            className="Searchbar"
            onKeyDown={[Function]}
          >
            <div
              className="Searchbar--Container"
            >
              <input
                autoComplete="off"
                id="search"
                name="search"
                onChange={[Function]}
                placeholder="search"
                type="text"
                value=""
              />
              <div
                className="Searchbar--Icon"
                onClick={[Function]}
              >
                <svg
                  height="18"
                  viewBox="0 0 18 18"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clipRule="evenodd"
                    d="M12.5 11H11.7L11.4 10.7C12.4 9.6 13 8.1 13 6.5C13 2.9 10.1 0 6.5 0C2.9 0 0 2.9 0 6.5C0 10.1 2.9 13 6.5 13C8.1 13 9.6 12.4 10.7 11.4L11 11.7V12.5L16 17.5L17.5 16L12.5 11ZM6.5 11C4 11 2 9 2 6.5C2 4 4 2 6.5 2C9 2 11 4 11 6.5C11 9 9 11 6.5 11Z"
                    fillRule="evenodd"
                  />
                </svg>
              </div>
            </div>
          </div>
        </Searchbar>
        <Subreddits>
          <section
            className="pt-12 Subreddits"
          >
            <h2
              className=""
            >
              subreddits
            </h2>
            <ul
              className="subreddits--list"
            />
          </section>
        </Subreddits>
      </div>
    </aside>
    <aside
      className="Mobile--Sidebar sm:flex hidden"
      data-mobile-menu={true}
      onClick={[Function]}
    >
      <span
        className="menu"
      >
        <HiOutlineMenuAlt2>
          <IconBase
            attr={
              Object {
                "fill": "none",
                "stroke": "currentColor",
                "viewBox": "0 0 24 24",
              }
            }
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6h16M4 12h16M4 18h7"
                key="0"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
              />
            </svg>
          </IconBase>
        </HiOutlineMenuAlt2>
         
      </span>
    </aside>
  </Sidebar>
</Provider>
`;
